Preserve Symlinks When Copying Additional Build Files

clickthisnick opened the pull request:
Fixes issue reported in https://github.com/ansible/ansible-builder/issues/749

clickthisnick committed:
Dont follow symlinks when copying files

clickthisnick committed:
Write tests for not following symlinks during copy files

Shrews commented on src/ansible_builder/utils.py:
```diff
@@ -200,7 +200,7 @@ def copy_file(source: str, dest: str, ignore_mtime: bool = False) -> bool:
 
     if should_copy:
-        shutil.copy2(source, dest)
+        shutil.copy2(source, dest, follow_symlinks=False)
```
One thing I noticed is that this will now error on subsequent runs of `create` because the symlink will already exist (does not happen on regular files).

clickthisnick replied:
ack good call - can fix that but wondering about something else now - if I remove a file and then run create twice it doesn't get removed from the context.

I usually just `rm -rf context` when I've used ansible-builder so I didn't even think of running it twice - but in this case it should remove the removed files from the context and the resulting image right?

Reproduction:
```bash
mkdir -p ansible/src
touch ansible/src/foo
cat <<EOF > ansible/execution-environment.yml
---
version: 3

dependencies:
  ansible_core:
    package_pip: ansible-core==2.18.6
  ansible_runner:
    package_pip: ansible-runner==2.4.1
  python_interpreter:
      package_system: "python312"
      python_path: "/usr/bin/python3.12"

images:
  base_image:
    name: quay.io/ansible/ansible-runner:latest

additional_build_files:
  - src: src
    dest: src

additional_build_steps:
  append_base: |
    COPY _build/src /src
EOF

cd ansible
ansible-builder create
ansible-builder build
docker run ansible-execution-env ls /src
> foo
rm src/foo
ansible-builder create
ansible-builder build
docker run ansible-execution-env ls /src
> foo (but I deleted foo create/build did not remove from the context)
touch src/apple
ansible-builder create
ansible-builder build
docker run ansible-execution-env ls /src
> apple (It added this to the context)
> foo (But never removed this)
```

clickthisnick replied:
this was a bit tricky because of how symlinks are handled including ones that point to files that don't exist but updated the code and unit tests to reflect what I think this function should do

I still don't quite grok that once a file is in the build context/cache running ansible-builder create won't remove it

Shrews reviewed (commented):
Besides the potential issue I pointed out, I/we need to give some thought on how your use case of using symlinks is going to interact with the code that attempts to preserve the build cache by checking file stats. I'm not certain if there is any impact there at all.

clickthisnick committed:
Handle symlinks

Fix when running ansible-builder create twice

If symlink changes we will override it
If symlink is not changed we will not copy

clickthisnick committed:
Handle all permutations of symlink copying

Copy same symlink
Copy different symlink
Copy broken symlink
Copy file over symlink
Copy symlink over file

clickthisnick committed:
Write unit tests for each symlink permutation

Shrews commented on src/ansible_builder/utils.py:
```diff
@@ -163,44 +163,65 @@ def copy_directory(source_dir: Path, dest: Path):
+    :param dest: Destination file path.
+    :param ignore_mtime: If True, modification times are ignored in change detection.
+    :returns: True if the file was copied, False otherwise.
+    :raises: Exception if source or destination is a directory.
```
Let's not reformat this.

Shrews commented on src/ansible_builder/utils.py:
```diff
@@ -163,44 +163,65 @@ def copy_directory(source_dir: Path, dest: Path):
-        logger.warning('File %s updated time increased and will be rewritten', dest)
+
+    if source_path.is_dir() or dest_path.is_dir():
+        raise Exception(f"{'Source' if source_path.is_dir() else 'Destination'} {source} can not be a directory. Please use copy_directory instead.")
```
This isn't correct in that it always references `{source}`. Generally, reformatting of code that isn't specific to your change is just "noise" that makes the code more difficult to review. I suggest returning it to the original format.

clickthisnick replied:
it works because of `if source_path.is_dir() or dest_path.is_dir():` but ack - will revert

Shrews commented on src/ansible_builder/utils.py:
```diff
@@ -163,44 +163,65 @@ def copy_directory(source_dir: Path, dest: Path):
+            logger.debug("Symlink %s target differs and will be overwritten.", dest)
+        elif dest_path.exists():
+            logger.debug("Destination %s is a regular file and will be replaced by symlink.", dest)
+
```
I'm wondering if maybe it would be better if, instead of comparing _what_ the symlinks point to, we just let the `filecmp.cmp()` call below take care of deciding if there are changes? I would think the contents would be more important than the actual file being pointed to. I _think_ with the original code, the `filecmp.cmp()` would have worked with symlinks as-is. Probably needs testing though.

clickthisnick replied:
`filecmp.cmp()` with symlinks make the unit tests indeterministic - I install pytest-repeat and ran them 1000 times and copied which I would expect to be False would switch between True and False.

I kept the reproduction steps in the unit tests, but added an extra case before `filecmp.cmp()` which explicitly checks for the symlink conditions

I couldn't track why `filecmp.cmp()` was doing this tho - I was thinking it was something in the os or the pytest fixture for creating tmp files but couldn't figure it out

Shrews commented on src/ansible_builder/utils.py:
```diff
@@ -163,44 +163,65 @@ def copy_directory(source_dir: Path, dest: Path):
     if should_copy:
-        shutil.copy2(source, dest)
+        if dest_path.is_symlink() or dest_path.exists():
+            os.unlink(dest)
```
No need to remove normal files here, just symlinks.

clickthisnick replied:
This was to handle the case where:

source = symlink
dest = file

when I run `ansible-builder create`  I expect dest to now be a symlink

that maybe untrue or too much of an edge case - I can guard on that specific case in order to not remove the file each time

Shrews commented on src/ansible_builder/utils.py:
```diff
@@ -163,44 +163,65 @@ def copy_directory(source_dir: Path, dest: Path):
+            os.symlink(os.readlink(source), dest)
+        else:
+            shutil.copy2(source, dest)
+
```
Could we not just set `follow_symlinks` in `copy2()` to the correct value based on if it is a symlink or not?

clickthisnick replied:
yea good point - will update - thanks for the comments

Shrews reviewed (changes requested):
Left you some comments inline. In general, unnecessary code or comment reformatting just adds noise for the reviewer, so it's best to avoid that when possible.

clickthisnick committed:
Address comments / Simplify logic while making unit tests deterministic

filecmp.cmp() has an issue with symlinks

I'm thinking its perhaps within the os or the pytest tmp_path fixture but I couldn't track it down

Using pytest-repeat --count=1000 you get different results 950~ True / 50~ False

I tried various things like ensuring flushing the files but couldn't make it deterministic with the filecmp.cmp() check

clickthisnick committed:
Add more unit tests with reproduction cases to indeterministic results

clickthisnick commented on test/unit/test_utils.py:
```diff
@@ -145,3 +157,179 @@ def test_copy_directory(tmp_path):
+    # 1000 runs using pytest-repeat
+    # 37 times it is False
+    # 963 times it is True
+    assert copied in (False, True)
```
I am confused by this and couldn't think of a way to get around it

Shrews replied:
That's very odd. I changed the assert to check for `False` and ran the test 5000 times and it never failed.

clickthisnick replied:
I'm developing within a container maybe that is what is going on - I'll try on mac osx directly

I'm fine with just changing the assertion to False, removing the pytest-repeat comments in the tests and knowing that my environment is weird and it might invalidate the cache for me

Does that seem okay given you ran it 5k times and it always worked 😅

Shrews replied:
Yes, definitely change it to the expected value, otherwise this isn't really a good test. Can remove all of the comments about the tests failing too. It likely has something to do with your environment. In the meantime, I'll test it some more on my other systems, just to verify..

clickthisnick replied:
Interesting - the CI run reproduced the same issue I was experiencing locally

clickthisnick replied:
It def wasn't deterministically failing for me - curious if the CI ran again if it would pass/fail on a different interpreter

Shrews replied:
I still cannot reproduce this on my systems. But, on a hunch, can you try adding a call to `filecmp.clear_cache()` in between the consecutive `copy_file()` calls and see if you still get the failures on your machine?

clickthisnick replied:
Your suggestion fixes things - I didn't realize it had a cache, was focusing on the files themselves. The outcome (False -> True) was cached is the issue:

After First Copy (Before Cache Clearing):
`
filecmp._cache:
{('/tmp/pytest-of-clickthisnick/pytest-122/test_regular_file_overwrites_s0/source.txt', '/tmp/pytest-of-clickthisnick/pytest-122/test_regular_file_overwrites_s0/dest_symlink', (32768, 3, 1751046892.509871), (32768, 3, 1751046892.509871)): False}
`

After Cache Clearing and Second Copy:
`
filecmp._cache:
{('/tmp/pytest-of-clickthisnick/pytest-122/test_regular_file_overwrites_s0/source.txt', '/tmp/pytest-of-clickthisnick/pytest-122/test_regular_file_overwrites_s0/dest_symlink', (32768, 3, 1751046892.509871), (32768, 3, 1751046892.509871)): True}
`

clickthisnick replied:
Do you have any preference what should happen here

If I call `ansible-builder create` twice there is no shared cache between the runs so that's okay
but if `copy_file` gets twice in the same program it could have this issue

-----

I'm thinking:

1.) In `copy_file` we could look if the input paths are already in cache and pop them (or just clear all cache) before using `filecmp.cmp()` to check
 
 2.) we could just add to the unit test which is probably fine for most current use-cases since each `ansible-builder create` has its own cache in memory

clickthisnick replied:
minimal reproduction case that works at least in my env
```python
import filecmp, shutil

with open("/tmp/foo", "w") as f:
    f.write("hello")

with open("/tmp/bar", "w") as f:
    f.write("world")

# Should be False
print(filecmp.cmp("/tmp/foo", "/tmp/bar", shallow=False))

# Overwrite bar with foo (same size)
shutil.copyfile("/tmp/foo", "/tmp/bar")

# Still returns False due to stale cache!
print(filecmp.cmp("/tmp/foo", "/tmp/bar", shallow=False))  # ❌ Incorrect

# Correct after clearing cache
filecmp.clear_cache()
print(filecmp.cmp("/tmp/foo", "/tmp/bar", shallow=False)) 
```

Shrews replied:
We don't need to modify `copy_file()` here. This only happens in the test because `filecmp.cmp()` is called back-to-back on the same file after it changes which, [according to the `filecmp` docs](https://docs.python.org/3/library/filecmp.html#filecmp.clear_cache), can cause cache issues. This isn't going to happen in the normal code. So just putting the `clear_cache()` call in the tests, with a comment on why we are adding it, is the proper solution, IMO.

clickthisnick committed:
Symlink check needs to go before checking if file exists

This is to capture the case where we have a symlink that points to a file that doesn't exist yet (It could be created a runtime via a mount or a local task)

clickthisnick committed:
Update tests to handle broken symlink case not invalidating cache 

Source needed to move the symlink check before the file exists check

clickthisnick commented on src/ansible_builder/utils.py:
```diff
@@ -189,7 +189,10 @@ def copy_file(source: str, dest: str, ignore_mtime: bool = False) -> bool:
     if Path(dest).is_dir():
         raise Exception(f"Destination {dest} can not be a directory. Please use copy_directory instead.")
-    if not os.path.exists(dest):
+    if Path(source).is_symlink() and Path(dest).is_symlink() and os.readlink(source) == os.readlink(dest):
```
I couldn't figure out why `not filecmp.cmp(source, dest, shallow=False)` wasn't working properly for symlinks so I just added this as an explicit check.

It needs to be before:
`elif not os.path.exists(dest):`

To handle the case where the symlink is pointing to a file that does not yet exist (on purpose)

Shrews replied:
`filecmp.cmp()` should totally work with symlinks. Can you show me an example where it isn't working?

Shrews replied:
Oh, maybe you meant `filecmp.cmp()` didn't work _because_ the link was pointing to a non-existing file? If so, then yeah, an exception would be thrown in that case since that function attempts to `stat` the actual file (not the link). I.e., it follows the symlink for comparing.

clickthisnick committed:
remove pytest comments, add expected copied value

clickthisnick committed:
address comment / fix unit test

Shrews reviewed (approved):
Thanks for the contribution.

Since this is a slight change in behavior, I think it is best suited for the next 3.2 release, and so we won't backport it for another 3.1.x release.
